### DISCUSSION TITLE
dirichlet boundary conditions in StationaryModel

### DISCUSSION THREAD

**Discussion by lebeand** [pymor/pymor#1292] [Q&A] [answered]
If there are dirichlet boundary conditions (dbc), the resulting discretization (with the discretization toolkit) contains the points for which the solution is already set through dbc. Is it possible to get a `StationaryModel` without these points? 

In this code below we have a symmetric, positive definite stiffness matrix. But with the dbc points, the matrix `A_with_dirichlet` is unsymmetric. If I would want to compute the coercivity constant for one parameter I would have to delete certain rows, columns manually and use `A` for solving a generalized eigenvalue problem. `A` is s.p.d how expected.

My code:
```python
# import libraries
import numpy as np
from pymor.basic import *
from matplotlib import pyplot as plt
import random

from pymor.parameters.functionals import MinThetaParameterFunctional
from scipy import linalg
import pickle


# define problem
domain = RectDomain(domain=([-1., -1.], [1., 1.]), top='dirichlet', left='neumann', right='neumann', bottom='neumann')

coeff = 1.
func = ExpressionFunction('ones(x.shape[:-1])', 2, ())

diffusion = -LincombFunction([func],
                            [coeff])

rhs = ConstantFunction(0, 2)

neumann_data = LincombFunction([ExpressionFunction('(x[...,1] <= -1.+1e-15) * 1.', 2, ())],
                               [ExpressionParameterFunctional('mu2[0]', {'mu2': 1})])

problem = StationaryProblem(domain=domain,
                            diffusion=diffusion,
                            rhs=rhs,
                            parameter_ranges={'mu2': (-1, 1)},
                            neumann_data=neumann_data,
                            )

# discretize problem
fom, fom_data = discretize_stationary_cg(problem, grid_type=TriaGrid, diameter=1/4)

# full order solution
mu2 = 0.94
mu = Mu({'mu2':mu2})
U_fom = fom.solve(mu)
U_fom_numpy=U_fom.to_numpy()


A_with_dirichlet=fom.operator.assemble(mu2).matrix.toarray()
print(np.max(A_with_dirichlet-A_with_dirichlet.T))

# search for dirichlet grid points in fom_data['grid'].centers(2) and remove the corresponding rows, columns
A=np.delete(A_with_dirichlet,np.array(range(72,81)),axis=0)
A=np.delete(A,np.array(range(72,81)),axis=1)
print(np.max(A-A.T))

'''
.
.
.
'''


```

**Comment by sdrave** [comment, marked as answer]
Hi @andre906, pyMOR's discretization toolkit does not allow eliminating Dirichlet DOFs from the solution space, unfortunately. In fact, I am not aware of any popular PDE solver which does this. (If you have an example, I would be very interested to know...) What you can do is clear the rows as well as the columns of the Dirichlet DOFs. In fact, [we already do this](https://github.com/pymor/pymor/blob/main/src/pymor/discretizers/builtin/cg.py#L1145) to obtain spd inner product matrices. You can also choose to remove the ones from the diagonals if you wish. Since the last release you can also specify `mu_energy_product` when calling `discretize_stationary_cg` to obtain an spd energy product operator.

If you really want to eliminate the DOFs, you can get all Dirichlet DOFs using `fom_data['boundary_info'].dirichlet_boundaries(2)`.  Some code to eliminate rows and columns from scipy sparse matrices can be found [here](https://stackoverflow.com/questions/13077527/).

**Reply by lebeand** [reply]
Okay, thank you!
